File: backend/app/models/payment.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class Payment(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "payments"

    clinic_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("clinics.id"), nullable=False, index=True
    )
    booking_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("bookings.id"), nullable=True
    )
    customer_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("customers.id"), nullable=False
    )

    # Payment type
    payment_type: Mapped[str] = mapped_column(String(20), nullable=False)
    # deposit / remaining / full / additional

    # Amount
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(5), default="KRW")

    # PG provider info
    pg_provider: Mapped[str | None] = mapped_column(String(50))
    pg_payment_id: Mapped[str | None] = mapped_column(String(200))
    payment_method: Mapped[str | None] = mapped_column(String(50))

    # Payment link / QR
    payment_link: Mapped[str | None] = mapped_column(String(500))
    qr_code_url: Mapped[str | None] = mapped_column(String(500))
    link_expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # Status
    status: Mapped[str] = mapped_column(String(20), default="pending")
    # pending / link_sent / completed / failed / refunded / cancelled

    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # Metadata
    metadata_: Mapped[dict] = mapped_column("metadata", JSONB, default=dict)

    # Relationships
    clinic: Mapped["Clinic"] = relationship()  # noqa: F821
    booking: Mapped["Booking | None"] = relationship()  # noqa: F821
    customer: Mapped["Customer"] = relationship()  # noqa: F821

    def __repr__(self) -> str:
        return f"<Payment {self.id} status={self.status}>"

File: frontend/src/types/payment.ts
export interface Payment {
  id: string;
  clinic_id: string;
  booking_id: string | null;
  customer_id: string;
  payment_type: string;
  amount: number;
  currency: string;
  pg_provider: string | null;
  pg_payment_id: string | null;
  payment_method: string | null;
  payment_link: string | null;
  qr_code_url: string | null;
  link_expires_at: string | null;
  status: string;
  paid_at: string | null;
  created_at: string;
  updated_at: string;
}

export type PaymentStatus =
  | "pending"
  | "completed"
  | "failed"
  | "refunded"
  | "expired";

export const PAYMENT_STATUS_LABELS: Record<PaymentStatus, string> = {
  pending: "대기",
  completed: "완료",
  failed: "실패",
  refunded: "환불",
  expired: "만료",
};

export const PAYMENT_STATUS_COLORS: Record<PaymentStatus, string> = {
  pending: "bg-yellow-50 text-yellow-700",
  completed: "bg-green-50 text-green-700",
  failed: "bg-red-50 text-red-700",
  refunded: "bg-purple-50 text-purple-700",
  expired: "bg-gray-100 text-gray-500",
};
